src/main/java/net/mcreator/ui/modgui/ModElementChangedListener.java

```diff
@@ -0,0 +1,139 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package net.mcreator.ui.modgui;
+
+import javafx.embed.swing.JFXPanel;
+import net.mcreator.ui.component.JItemListField;
+import net.mcreator.ui.minecraft.JEntriesList;
+import net.mcreator.ui.minecraft.MCItemHolder;
+
+import javax.swing.*;
+import javax.swing.event.ChangeEvent;
+import javax.swing.event.ChangeListener;
+import javax.swing.event.DocumentEvent;
+import javax.swing.event.DocumentListener;
+import javax.swing.text.JTextComponent;
+import java.awt.*;
+import java.awt.event.*;
+
+/**
+ * <p>Gets triggered whenever a change to a mod element is detected inside {@link ModElementGUI}</p>
+ *
+ * <p>One case of this listener usage is warning the user about some unsaved changes made to a mod element.
+ * Another case is the {@link net.mcreator.ui.modgui.codeviewer.ModElementCodeViewer ModElementCodeViewer} which
+ * regenerates the code preview upon UI changes.</p>
+ */
+public interface ModElementChangedListener
+		extends MouseListener, KeyListener, ActionListener, ChangeListener, DocumentListener {
+
+	/**
+	 * <p>The main listener method, triggered when an event occurs on a registered container</p>
+	 */
+	void modElementChanged();
+
+	/**
+	 * <p>Registers the given UI component to trigger this listener when a change is detected on it</p>
+	 *
+	 * @param container The UI element to register
+	 */
+	default void registerUI(JComponent container) {
+		for (Component component : container.getComponents()) {
+			if (component instanceof MCItemHolder itemHolder) {
+				itemHolder.addBlockSelectedListener(this);
+			} else if (component instanceof JItemListField<?> listField) {
+				listField.addChangeListener(this);
+			} else if (component instanceof JEntriesList entriesList) {
+				registerUI(entriesList);
+				entriesList.addEntryRegisterListener(c -> {
+					registerUI(c);
+					modElementChanged();
+				});
+				component.addMouseListener(this);
+			} else if (component instanceof AbstractButton button) {
+				button.addActionListener(this);
+			} else if (component instanceof JSpinner spinner) {
+				spinner.addChangeListener(this);
+			} else if (component instanceof JComboBox<?> comboBox) {
+				comboBox.addActionListener(this);
+			} else if (component instanceof JTextComponent textComponent) {
+				textComponent.getDocument().addDocumentListener(this);
+			} else if (component instanceof JFXPanel) {
+				component.addMouseListener(this);
+				component.addKeyListener(this);
+			} else if (component instanceof JComponent jcomponent) {
+				registerUI(jcomponent);
+
+				if (!(component instanceof JLabel) && !(component instanceof JPanel)

```

Comment:
At the MCreator's "taskbar" or in the code viewer?

Reply:
Categories, like bounding boxes, advanced properties, triggers, ...

Reply:
Those are driven by JButtons, so to avoid 'overkills', I need to exclude `JModElementProgressPanel` from the `centerComponent`.
Or make a special button type.